---
title: "dtuSimulationsSoneson"
author: "Koen Van den Berge"
date: "3 January 2017"
output: html_document
---

# Simulations from Soneson F1000 paper.

```{r}
library(tximport)
library(edgeR)
library(iCOBRA)
library(dplyr)
library(Hmisc)
library(BiocParallel)
library(DESeq2)
library(DEXSeq)
library(biomaRt)
library(Biostrings)
library(scales)
source("/Users/koenvandenberge/Dropbox/PhD/Research/transcriptLevel/helpFunctionsCharlotte.R")
muted <- c("#DC050C","#E8601C","#7BAFDE","#1965B0","#B17BA6",
           "#882E72","#F1932D","#F6C141","#F7EE55","#4EB265",
           "#90C987","#CAEDAB","#777777")
meta <- data.frame(sample = paste0("sample", c("A1", "A2", "A3", "B1", "B2", "B3")),
                   condition = c("A", "A", "A", "B", "B", "B"),
                   stringsAsFactors = FALSE)
rownames(meta) <- meta$sample


### get GTF file
#library(rtracklayer)
## download the GTF file and save it.
#gtfWholeGenome="/Users/koenvandenberge/PhD_Data/dtu/annotation/Homo_sapiens.GRCh37.71.gtf"
#gtff <- import(gtfWholeGenome)
#gtff2 <- subset(gtff, seqnames == "1")
#export(gtff2, "/Users/koenvandenberge/PhD_Data/dtu/annotation/Homo_sapiens.GRCh37.71.chr1.gtf", format = "gtf")

## generate tx 2 gene map
cdna_fasta <-"/Users/koenvandenberge/Dropbox/edgeR_zeroinflation/simulated_data_bulk_tx/sim2_human/reference_files/Homo_sapiens.GRCh37.71.cdna.chr1.fa"
gtf <- "/Users/koenvandenberge/PhD_Data/dtu/annotation/Homo_sapiens.GRCh37.71.chr1.gtf"
feature_lengths_file <- "/Users/koenvandenberge/PhD_Data/dtu/annotation/feature_lengths.Rdata"
tx_gene_file <- "/Users/koenvandenberge/PhD_Data/dtu/annotation/tx_gene_map.Rdata"
#calc_lengths_mapping(gtf = gtf, cdna_fasta = cdna_fasta, 
#                       feature_lengths_file = feature_lengths_file, 
#                       tx_gene_file = tx_gene_file) 
load(feature_lengths_file)
load(tx_gene_file) #gene2tx
tx2gene = gene2tx[,c(2,1)]

truth_tx_file <- "/Users/koenvandenberge/Dropbox/PhD/Research/stageWiseTesting/githubPaper/stageWiseTestingPaper/DTU_DTE/simulation/introduction/truth_transcript.txt"
truth_tx <- read.delim(truth_tx_file, header = TRUE, as.is = TRUE)


### derive salmon transcript counts
salmon_basedir <- "/Users/koenvandenberge/PhD_Data/dtu/sim2_human/salmon"

salmon_files <- list.files(salmon_basedir, pattern = "sample", full.names = TRUE)
salmon_files <- salmon_files[file.info(salmon_files)$isdir]
salmon_files <- paste0(salmon_files, "/quant.sf")
salmon_files <- salmon_files[file.exists(salmon_files)]
names(salmon_files) <- basename(gsub("/quant.sf", "", salmon_files))
txi_salmonsimplesum <- tximport(files = salmon_files, type = "salmon", txIn = TRUE,
                                txOut = FALSE, countsFromAbundance = "no", 
                                tx2gene = tx2gene)

txi_salmonscaledtpm <- tximport(files = salmon_files, type = "salmon", txIn = TRUE,
                                txOut = FALSE, countsFromAbundance = "scaledTPM", 
                                 tx2gene = tx2gene)

txi_salmontx <- tximport(files = salmon_files, type = "salmon", txIn = TRUE,
                         txOut = TRUE, countsFromAbundance = "no", tx2gene = tx2gene)

salmon_quant <- list(geneCOUNT_sal_simplesum = txi_salmonsimplesum$counts,
                     geneCOUNT_sal_scaledTPM = txi_salmonscaledtpm$counts,
                     avetxlength = txi_salmonsimplesum$length,
                     geneTPM_sal = txi_salmonsimplesum$abundance,
                     txTPM_sal = txi_salmontx$abundance,
                     txCOUNT_sal = txi_salmontx$counts,
                     txi_salmonsimplesum = txi_salmonsimplesum,
                     txi_salmonscaledtpm = txi_salmonscaledtpm,
                     txi_salmontx = txi_salmontx)
stopifnot(all(colnames(salmon_quant$txCOUNT_sal) == rownames(meta)))

#### DEXSEQ on salmon transcript counts
## discard genes with only one transcript and not expressed transcripts
txCount <- ceiling(salmon_quant$txCOUNT_sal)
txCount <- txCount[!rowSums(txCount)==0,]
genesWithOneTx <- names(table(tx2gene$gene))[table(tx2gene$gene)==1]
txFromGenesWithOneTx <- tx2gene$transcript[match(genesWithOneTx,tx2gene$gene)]
txCount <- txCount[!rownames(txCount)%in%txFromGenesWithOneTx,]

## regular DEXSeq analysis
dxd_sal <- DEXSeqDataSet(countData = txCount, 
                         sampleData = meta, 
                         design = ~sample + exon + condition:exon,
                         featureID = rownames(txCount),
                         groupID = tx2gene$gene[match(rownames(txCount),
                                                      tx2gene$transcript)])
dxd_sal <- estimateSizeFactors(dxd_sal)
dxd_sal <- estimateDispersions(dxd_sal)
dxd_sal <- testForDEU(dxd_sal)
dxr_sal <- DEXSeqResults(dxd_sal)
hist(dxr_sal$pvalue)
qval_dtu_salmon <- perGeneQValue(dxr_sal)

## ROC gene-level analysis
truth_gene_file <- "/Users/koenvandenberge/Dropbox/PhD/Research/stageWiseTesting/githubPaper/stageWiseTestingPaper/DTU_DTE/simulation/introduction/truth_gene.txt"
truth_gene <- read.delim(truth_gene_file, header = TRUE, as.is = TRUE, row.names = 1)
cobra <- COBRAData(padj = data.frame(salmon_dexseq = qval_dtu_salmon,
                                     row.names = names(qval_dtu_salmon),
                                     stringsAsFactors = FALSE))
cobra <- COBRAData(truth = truth_gene, object_to_extend = cobra)
cobraperf <- calculate_performance(cobra, binary_truth = "diffisouse")
cobraplotGene <- prepare_data_for_plot(cobraperf, incltruth = TRUE, 
                                   colorscheme = c("blue", "red", "green","steelblue"),
                                   facetted = FALSE)
plot_fdrtprcurve(cobraplotGene)
plot_roc(cobraplotGene,xaxisrange=c(0,0.1))

## ROC transcript-level analysis
rownames(truth_tx) <- truth_tx$transcript
cobra <- COBRAData(padj = data.frame(salmon_dexseq = p.adjust(dxr_sal$pvalue,"BH"),
                                     row.names = dxr_sal$featureID,
                                     stringsAsFactors = FALSE))
cobra <- COBRAData(truth = truth_tx, object_to_extend = cobra)
cobraperf <- calculate_performance(cobra, binary_truth = "status")
cobraplotTx <- prepare_data_for_plot(cobraperf, incltruth = TRUE, 
                                   colorscheme = c("blue", "red", "green","steelblue"),
                                   facetted = FALSE)
plot_fdrtprcurve(cobraplotTx)
plot_roc(cobraplotTx,xaxisrange=c(0,0.1))


### combine in one plot
library(scales)
par(bty="l", cex.axis=1.5, cex.lab=1.5, mar=c(5,4.5,4,1))
plot(x=cobraplotGene@fdrtprcurve$FDR,y=cobraplotGene@fdrtprcurve$TPR, type="l",col="black",lwd=2, xlim=c(0,0.7), ylab="True Positive Rate", xlab="False Discovery Proportion")
points(x=cobraplotGene@fdrtpr$FDR,y=cobraplotGene@fdrtpr$TPR, col="white", pch=19, cex=1.2)
points(x=cobraplotGene@fdrtpr$FDR,y=cobraplotGene@fdrtpr$TPR, col="black", pch="o", cex=1.2)
lines(x=cobraplotTx@fdrtprcurve$FDR,y=cobraplotTx@fdrtprcurve$TPR,col="red",lwd=2)
points(x=cobraplotTx@fdrtpr$FDR,y=cobraplotTx@fdrtpr$TPR, col="white", pch=19, cex=1.2)
points(x=cobraplotTx@fdrtpr$FDR,y=cobraplotTx@fdrtpr$TPR, col="red", pch="o", cex=1.2)
abline(v=c(0.01,0.05,0.1,seq(0.1,1,.1)), col=alpha("grey",.8), lty=2)
legend("bottomright",c("gene-level","transcript-level"),col=c("black","red"),lty=1, bty="n", cex=1.5)

cobraPlotGeneF1000=cobraplotGene
cobraPlotTxF1000=cobraplotTx
```

# Soneson Genome Biology simulations

```{r}
### simulation Drosophila from Genome Biology paper Soneson 2016
baseDir <- "/Users/koenvandenberge/PhD_Data/dtu/E-MTAB-3766/diff_splicing_comparison_drosophila/"
files=list.files(baseDir)
fastaFiles <- files[grep(x=files,pattern=".fq.gz")]
meta=read.delim("/Users/koenvandenberge/PhD_Data/dtu/E-MTAB-3766/E-MTAB-3766.sdrf.txt",header=TRUE)
names(fastaFiles) <- (meta$Array.Data.File[1:12])
kallistoIndex="/Users/koenvandenberge/PhD_Data/dtu/diff_splice_paper_Kvdb/drosophila/reference_files/KallistoIndex/Drosophila_melanogaster.BDGP5.70.dna.toplevel"
txConversionFile="/Users/koenvandenberge/PhD_Data/dtu/diff_splice_paper_Kvdb/drosophila/reference_files/KallistoIndex/TranscriptID_conversion.txt"
kallistoDir="/Users/koenvandenberge/PhD_Data/dtu/E-MTAB-3766/diff_splicing_comparison_drosophila/quantifications/kallisto/"
truthFile="/Users/koenvandenberge/PhD_Data/dtu/E-MTAB-3766/Dm_truth.txt"
library(DEXSeq)
library(iCOBRA)
library(scales)
library(dplyr)

### kallisto quantification
sample <- unlist(lapply(strsplit(names(fastaFiles),split="_"),function(x) x[3]))
fileNames <- unlist(lapply(strsplit(names(fastaFiles),split=".",fixed=TRUE),function(x) x[1]))
fileNames <- substr(x=fileNames,1,11)[seq(1,12,2)]
#for(i in 1:6){
#    pairedFasta <- fastaFiles[sample==i]
#    cmd <- paste0("kallisto quant -i ",kallistoIndex,
#	      " -o ",kallistoDir,fileNames[i],
#	      " -b 30 ",
#	      baseDir,pairedFasta[1]," ",baseDir,pairedFasta[2])
#    message(cmd)
#    system(cmd)
#}

### get kallisto results
files2 <- list.files(kallistoDir)
sampleDirs <- files2[grep(x=files2,pattern="Dm_sample_[1-6]")]
dir=sampleDirs[1]
hlp=read.table(paste0(kallistoDir,"/",dir,"/abundance.tsv"), header=TRUE)
data <- as.data.frame(sapply(sampleDirs,function(dir) read.table(paste0(kallistoDir,"/",dir,"/abundance.tsv"), header=TRUE)[,"est_counts"]), row.names=hlp$target_id)
kal2tx=read.table(txConversionFile)
colnames(kal2tx) <- c("kallisto","transcript")
rownames(data) <- kal2tx$transcript[match(kal2tx$kallisto,rownames(data))]

## the truth_tx file is incorrect for tx level evaluation: all tx from a gene get a differential splicing status, but actually only two of them should have.
truth <- read.table(truthFile,header=TRUE)
truth_gene <- truth[,c(1,3)]
truth_gene <- truth_gene[!duplicated(truth$gene_id),]
rownames(truth_gene) <- truth_gene[,1]

truth_tx <- truth[,2:3]
truth_tx$transcript_id <- as.character(truth_tx$transcript_id)
rownames(truth_tx) <- truth_tx[,1]

tx2gene <- truth[,1:2]
tx2gene$gene_id <- as.character(tx2gene$gene_id)
tx2gene$transcript_id <- as.character(tx2gene$transcript_id)

### check 2 most abundant tx
simFilesDir <- "/Users/koenvandenberge/PhD_Data/dtu/E-MTAB-3766/"
simFiles=list.files(simFilesDir)
firstSampleFiles <- simFiles[grep(x=simFiles,pattern="Dm_sample[1-3]")]
sample1Tpm <- read.table(paste0(simFilesDir,firstSampleFiles[1]),header=TRUE)
sample2Tpm <- read.table(paste0(simFilesDir,firstSampleFiles[2]),header=TRUE)
sample3Tpm <- read.table(paste0(simFilesDir,firstSampleFiles[3]),header=TRUE)

group1Tpm <- sample1Tpm
group1Tpm$TPM2 <- sample2Tpm$TPM
group1Tpm$TPM3 <- sample3Tpm$TPM
genesWithOneTx <- names(table(sample1Tpm$gene_id)[table(sample1Tpm$gene_id)==1])
group1Tpm <- group1Tpm[!group1Tpm$gene_id%in%genesWithOneTx,]
hlp=group_by(group1Tpm,by=gene_id)
highestExpressedTxPerGene <- dplyr::summarize(hlp, max1=transcript_id[order(apply(cbind(TPM,TPM2,TPM3),1,mean),decreasing=TRUE)[1]], max2=transcript_id[order(apply(cbind(TPM,TPM2,TPM3),1,mean),decreasing=TRUE)[2]])
head(highestExpressedTxPerGene)

truthTx <- data.frame(transcript=truth_tx[,1],tx_ds_status=0,row.names=rownames(truth_tx))
simulatedDTUGenes <- rownames(truth_gene[truth_gene$gene_ds_status==1,])
length(simulatedDTUGenes) #should be 1000
highestExpressedTxForSignificantGenes <- highestExpressedTxPerGene[highestExpressedTxPerGene$by%in%simulatedDTUGenes,]
dtuTx <- as.character(unlist(c(as.data.frame(highestExpressedTxForSignificantGenes)[,2:3])))
truthTx[dtuTx,"tx_ds_status"]=1


### DEXSeq analysis
genesWithOneTx <- names(table(tx2gene$gene))[table(tx2gene$gene)==1]
txFromGenesWithOneTx <- tx2gene$transcript[match(genesWithOneTx,tx2gene$gene)]
txCount <- ceiling(data)
txCount <- txCount[!rownames(txCount)%in%txFromGenesWithOneTx,]
txCount <- txCount[!rowSums(txCount)==0,]

geneTx <- tx2gene$gene_id[match(rownames(txCount),tx2gene$transcript_id)]
sampleData <- data.frame(condition=factor(rep(c("A","B"),each=3)))
dxd <- DEXSeqDataSet(countData = txCount, 
                         sampleData = sampleData, 
                         design = ~ sample + exon + condition:exon,
                         featureID = rownames(txCount),
                         groupID = geneTx)
dxd <- estimateSizeFactors(dxd)
dxd <- estimateDispersions(dxd)
dxd <- testForDEU(dxd)
dxr <- DEXSeqResults(dxd)
qvalDxr <- perGeneQValue(dxr)


### gene-level
truth_gene <- truth[,c(1,3)]
truth_gene <- truth_gene[!duplicated(truth$gene_id),]
rownames(truth_gene) <- truth_gene[,1]
cobra <- COBRAData(padj = data.frame(kallisto_dexseq = qvalDxr,
                                     row.names = names(qvalDxr),
                                     stringsAsFactors = FALSE))
cobra <- COBRAData(truth = truth_gene, object_to_extend = cobra)
cobraperf <- calculate_performance(cobra, binary_truth = "gene_ds_status")
cobraplotGene <- prepare_data_for_plot(cobraperf, incltruth = TRUE, 
                                   colorscheme = c("blue", "red", "green"),
                                   facetted = FALSE)
plot_fdrtprcurve(cobraplotGene)
plot_roc(cobraplotGene)

### transcript-level
padjTxDexSeq <- p.adjust(dxr$pvalue,"BH")
truth_tx <- truth[,2:3]
truth_tx$transcript_id <- as.character(truth_tx$transcript_id)
rownames(truth_tx) <- truth_tx[,1]
cobra <- COBRAData(padj = data.frame(kallisto_dexseq = padjTxDexSeq,
                                     row.names = dxr$featureID,
                                     stringsAsFactors = FALSE))
cobra <- COBRAData(truth = truthTx, object_to_extend = cobra)
cobraperf <- calculate_performance(cobra, binary_truth = "tx_ds_status")
cobraplotTx <- prepare_data_for_plot(cobraperf, incltruth = TRUE, 
                                   colorscheme = c("blue", "red", "green"),
                                   facetted = FALSE)
plot_fdrtprcurve(cobraplotTx)
plot_roc(cobraplotTx)

### combine in one plot
library(scales)
par(bty="l", cex.axis=1.5, cex.lab=1.5, mar=c(5,4.5,4,1))
plot(x=cobraplotGene@fdrtprcurve$FDR,y=cobraplotGene@fdrtprcurve$TPR,col=2,lwd=1, xlim=c(0,0.8), ylab="True Positive Rate", xlab="False Discovery Proportion", type="n")
abline(v=c(0.01,0.05,0.1,seq(0.1,1,.1)), col=alpha("grey",.8), lty=2)
lines(x=cobraplotGene@fdrtprcurve$FDR,y=cobraplotGene@fdrtprcurve$TPR,col="black",lwd=2)
points(x=cobraplotGene@fdrtpr$FDR,y=cobraplotGene@fdrtpr$TPR, col="white", pch=19, cex=1.2)
points(x=cobraplotGene@fdrtpr$FDR,y=cobraplotGene@fdrtpr$TPR, col="black", pch="o", cex=1.2)
lines(x=cobraplotTx@fdrtprcurve$FDR,y=cobraplotTx@fdrtprcurve$TPR,col="red",lwd=2)
points(x=cobraplotTx@fdrtpr$FDR,y=cobraplotTx@fdrtpr$TPR, col="white", pch=19, cex=1.2)
points(x=cobraplotTx@fdrtpr$FDR,y=cobraplotTx@fdrtpr$TPR, col="red", pch="o", cex=1.2)
legend("bottomright",c("gene-level","transcript-level"),col=c("black","red"),lty=1, bty="n", cex=1.5)
```


# Combine both performance plots

```{r}
par(mfrow=c(1,2))
#F1000
par(bty="l", cex.axis=1.5, cex.lab=1.5, mar=c(5,4.5,4,1))
plot(x=cobraPlotGeneF1000@fdrtprcurve$FDR,y=cobraPlotGeneF1000@fdrtprcurve$TPR, type="l",col="black",lwd=2, xlim=c(0,0.7), ylab="True Positive Rate", xlab="False Discovery Proportion")
points(x=cobraPlotGeneF1000@fdrtpr$FDR,y=cobraPlotGeneF1000@fdrtpr$TPR, col="white", pch=19, cex=1.2)
points(x=cobraPlotGeneF1000@fdrtpr$FDR,y=cobraPlotGeneF1000@fdrtpr$TPR, col="black", pch="o", cex=1.2)
lines(x=cobraPlotTxF1000@fdrtprcurve$FDR,y=cobraPlotTxF1000@fdrtprcurve$TPR,col="red",lwd=2)
points(x=cobraPlotTxF1000@fdrtpr$FDR,y=cobraPlotTxF1000@fdrtpr$TPR, col="white", pch=19, cex=1.2)
points(x=cobraPlotTxF1000@fdrtpr$FDR,y=cobraPlotTxF1000@fdrtpr$TPR, col="red", pch="o", cex=1.2)
abline(v=c(0.01,0.05,0.1,seq(0.1,1,.1)), col=alpha("grey",.8), lty=2)
legend("bottomright",c("gene-level","transcript-level"),col=c("black","red"),lty=1, bty="n", cex=1.5)

par(bty="l", cex.axis=1.5, cex.lab=1.5, mar=c(5,4.5,4,1))
plot(x=cobraplotGene@fdrtprcurve$FDR,y=cobraplotGene@fdrtprcurve$TPR,col=2,lwd=1, xlim=c(0,0.85), ylab="True Positive Rate", xlab="False Discovery Proportion", type="n")
abline(v=c(0.01,0.05,0.1,seq(0.1,1,.1)), col=alpha("grey",.8), lty=2)
lines(x=cobraplotGene@fdrtprcurve$FDR,y=cobraplotGene@fdrtprcurve$TPR,col="black",lwd=2)
points(x=cobraplotGene@fdrtpr$FDR,y=cobraplotGene@fdrtpr$TPR, col="white", pch=19, cex=1.2)
points(x=cobraplotGene@fdrtpr$FDR,y=cobraplotGene@fdrtpr$TPR, col="black", pch="o", cex=1.2)
lines(x=cobraplotTx@fdrtprcurve$FDR,y=cobraplotTx@fdrtprcurve$TPR,col="red",lwd=2)
points(x=cobraplotTx@fdrtpr$FDR,y=cobraplotTx@fdrtpr$TPR, col="white", pch=19, cex=1.2)
points(x=cobraplotTx@fdrtpr$FDR,y=cobraplotTx@fdrtpr$TPR, col="red", pch="o", cex=1.2)
legend("bottomright",c("gene-level","transcript-level"),col=c("black","red"),lty=1, bty="n", cex=1.5)
```

